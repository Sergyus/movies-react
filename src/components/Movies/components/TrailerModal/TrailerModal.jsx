import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import styles from './TrailerModal.scss';
import Spinner from '../../../Spinner';
import Modal from './components/Modal';
import { useBodyScrollLock, useOnClickOutside } from '../../../../utils';

const TrailerModal = ({ toggleModal, videoKey, isLoading }) => {
  useBodyScrollLock();
  let modal;
  const modalRef = useRef();
  useOnClickOutside(modalRef, () => toggleModal(false));

  if (isLoading) {
    modal = <Spinner />;
  } else {
    modal = (
      <>
        <iframe
          className={styles.video}
          title="trailer"
          frameBorder="0"
          src={`https://www.youtube.com/embed/${videoKey}`}
        />
      </>
    );
  }

  return (
    <Modal id="trailer" ref={modalRef}>
      <div className={styles.videoContainer}>
        {
          videoKey || isLoading
            ? modal
            : (
              <>
                <h2 className={styles.heading}>Sorry, no trailer available</h2>
              </>
            )
        }
      </div>
    </Modal>
  );
};

TrailerModal.propTypes = {
  videoKey: PropTypes.string,
  toggleModal: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
};

TrailerModal.defaultProps = {
  videoKey: '',
  isLoading: false,
};

export default TrailerModal;
