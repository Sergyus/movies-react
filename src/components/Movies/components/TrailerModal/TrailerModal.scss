@import "../../../../styles/variables.scss";

.videoContainer {
  position: fixed;
  z-index: 10;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding: 2.5em 1.5em 1.5em 1.5em;
  background-color: rgba(#000, 0.7);
  overflow-y: auto;
  -webkit-overflow-scrolling: touch;
  display: flex;
  justify-content: center;
  align-items: center;
}

.video {
  width: 50%;
  height: 70%;
  box-shadow: 0 0 10px 3px rgba(0, 0, 0, 0.9);
}

.heading {
  color: #fff;
}

@media (max-width: $desktop-screen-size) {
  .videoContainer {
    padding: 0;
  }

  .video {
    width: 90%;
  }
}
