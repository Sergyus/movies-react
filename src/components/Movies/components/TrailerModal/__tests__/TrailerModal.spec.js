import React from 'react';
import { create, act } from 'react-test-renderer';
import { JSDOM } from 'jsdom';
import ReactDOM from 'react-dom';
import TrailerModal from '..';

describe('TrailerModal component renders correctly', () => {
  beforeAll(() => {
    const dom = new JSDOM();
    global.document = dom.window.document;
    global.window = dom.window;
    ReactDOM.createPortal = jest.fn(element => element);
  });
  afterEach(() => {
    ReactDOM.createPortal.mockClear();
  });
  it('TrailerModal renders and unmount correctly', () => {
    const tree = create(<TrailerModal toggleModal={() => {}} />);
    expect(tree.toJSON()).toMatchSnapshot();
    tree.unmount();
  });
  it('Spinner inside modal renders correctly', () => {
    const tree = create(<TrailerModal toggleModal={() => {}} isLoading />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('Click outside modal works correctly', () => {
    const mockFn = jest.fn();
    const mockClick = jest.fn();
    const ContainerComponent = () => (
      <div>
        <button type="button" className="button" onClick={mockClick}>123</button>
        <TrailerModal toggleModal={mockFn} videoKey="12345" />
      </div>
    );
    let tree;
    act(() => {
      tree = create(
        <ContainerComponent />,
      );
    });
    const event = new MouseEvent('mousedown', {});
    act(() => {
      mockClick();
      document.dispatchEvent(event);
    });
    expect(mockClick).toHaveBeenCalled();
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
