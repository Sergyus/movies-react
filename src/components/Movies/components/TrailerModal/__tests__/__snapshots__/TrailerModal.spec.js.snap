// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrailerModal component renders correctly Click outside modal works correctly 1`] = `
<div>
  <button
    className="button"
    onClick={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    type="button"
  >
    123
  </button>
  <div
    className="videoContainer"
  >
    <iframe
      className="video"
      frameBorder="0"
      src="https://www.youtube.com/embed/12345"
      title="trailer"
    />
  </div>
</div>
`;

exports[`TrailerModal component renders correctly Spinner inside modal renders correctly 1`] = `
<div
  className="videoContainer"
>
  <div
    className="spinner"
  >
    <div
      className="ellipsis"
    >
      <div />
      <div />
      <div />
      <div />
    </div>
    <p
      className="title"
    >
      Loading
    </p>
  </div>
</div>
`;

exports[`TrailerModal component renders correctly TrailerModal renders and unmount correctly 1`] = `
<div
  className="videoContainer"
>
  <h2
    className="heading"
  >
    Sorry, no trailer available
  </h2>
</div>
`;
